aws-api/src/main/java/com/amplifyframework/api/aws/auth/ApiRequestSigner.java

```diff
@@ -0,0 +1,137 @@
+ * method.
+ */
+public abstract class ApiRequestSigner {
+    private static final String CONTENT_TYPE = "application/json";
+    private static final MediaType JSON_MEDIA_TYPE = MediaType.parse(CONTENT_TYPE);
+    private static final String APP_SYNC_SERVICE_NAME = "appsync";
+    private static final String API_GATEWAY_SERVICE_NAME = "apigateway";
+
+    abstract void addAuthHeader(com.amazonaws.Request<?> request);
+
+    /**
+     * Adds the appropriate headers to the provided HTTP request.
+     * @param req The request to be signed.
+     * @return A new instance of the request containing the signature headers.
+     * @throws IOException If the signing process fails.
+     */
+    public final Request sign(Request req) throws IOException {
+        //Clone the request into a new DefaultRequest object and populate it with credentials
+        final DefaultRequest<?> dr = new DefaultRequest<>(APP_SYNC_SERVICE_NAME);
+        //set the endpoint
+        dr.setEndpoint(req.url().uri());
+        //copy all the headers
+        for (String headerName : req.headers().names()) {
+            dr.addHeader(headerName, req.header(headerName));
+        }
+        //set the http method
+        dr.setHttpMethod(HttpMethodName.valueOf(req.method()));
+
+        //set the request body
+        final byte[] bodyBytes;
+        RequestBody body = req.body();
+        if (body != null) {
+            //write the body to a byte array.
+            final Buffer buffer = new Buffer();
+            body.writeTo(buffer);
+            bodyBytes = IOUtils.toByteArray(buffer.inputStream());
+        } else {
+            bodyBytes = "".getBytes();
+        }
+        dr.setContent(new ByteArrayInputStream(bodyBytes));
+
+        //set the query string parameters
+        dr.setParameters(splitQuery(req.url().url()));
+
+        addAuthHeader(dr);
+        //dr.addHeader(AUTHORIZATION, cognitoUserPoolsAuthProvider.getLatestAuthToken());
+        //dr.addHeader(AUTHORIZATION, oidcAuthProvider.getLatestAuthToken());
+
+        //Copy the signed/credentialed request back into an OKHTTP Request object.
+        Request.Builder okReqBuilder = new Request.Builder();
+
+        //set the headers from default request, since it contains the signed headers as well.
+        for (Map.Entry<String, String> e : dr.getHeaders().entrySet()) {
+            okReqBuilder.addHeader(e.getKey(), e.getValue());
+        }
+
+        //Set the URL and Method
+        okReqBuilder.url(req.url());
+        final RequestBody requestBody;
+        if (req.body() != null && req.body().contentLength() > 0) {
+            requestBody = RequestBody.create(bodyBytes, JSON_MEDIA_TYPE);
+        } else {
+            requestBody = null;
+        }
+
+        okReqBuilder.method(req.method(), requestBody);
+
+        //continue with chain.
+        return okReqBuilder.build();
+    }
+
+    // Extracts query string parameters from a URL.
```

Comment:
I realize this is existing code that you are just moving around, but it would be cleaner to use the `Uri` class in Android as [described here](https://stackoverflow.com/a/40466295/282279), instead of manually splitting the url string.

Reply:
Makes sense. I'll take a look at that. There are some other changes that need to be made in this code, so I'll be working on that today.